eland/ml/ltr/ltr_model_config.py

```diff
@@ -0,0 +1,153 @@
+        return {
+            self.type: {
+                k: v.to_dict() if hasattr(v, "to_dict") else v
+                for k, v in self.__dict__.items()
+                if v is not None and k != "type"
+            }
+        }
+
+
+class QueryFeatureExtractor(FeatureExtractor):
+    """
+    A class that allows to define a query feature extractor.
+    """
+
+    def __init__(
+        self,
+        feature_name: str,
+        query: Mapping[str, Any],
+        default_score: Optional[float] = float(0),
+    ):
+        """
+        Parameters
+        ----------
+        feature_name: str
+            Name of the extracted features.
+
+        query: Mapping[str, Any]
+            Templated query used to extract the feature.
+
+        default_score: str
+            Scored used by default when the doc is not matching the query.
+
+        Examples
+        --------
+        >>> from eland.ml.ltr import QueryFeatureExtractor
+
+        >>> query_feature_extractor = QueryFeatureExtractor(
+        ...    feature_name='title_bm25',
+        ...    query={ "match": { "title": "{{query}}" } }
+        ... )
+        """
+        super().__init__(feature_name=feature_name, type="query_extractor")
+        self.query = query
+        self.default_score = default_score
+
+
+class LTRModelConfig:
+    """
+    A class representing LTR model configuration.
+    """
+
+    def __init__(self, feature_extractors: List[FeatureExtractor]):
+        """
+        Parameters
+        ----------
+        feature_extractors: List[FeatureExtractor]
+            List of the feature extractors for the LTR model.
+
+        Examples
+        --------
+        >>> from eland.ml.ltr import LTRModelConfig, QueryFeatureExtractor
+
+        >>> ltr_model_config = LTRModelConfig(
+        ...     feature_extractors=[
+        ...        QueryFeatureExtractor(
+        ...            feature_name='title_bm25',
+        ...            query={ "match": { "title": "{{query}}" } }
+        ...        ),
+        ...        QueryFeatureExtractor(
+        ...            feature_name='descritption_bm25',
+        ...            query={ "match": { "description": "{{query}}" } }
+        ...        )
+        ...     ]
+        ... )
+        """
+        self.feature_extractors = feature_extractors
+
+    def to_dict(self) -> Mapping[str, Any]:
+        """
+        Convert the into a dict that can be send to ES as an inference config.
+        """
+        return {
+            TYPE_LEARNING_TO_RANK: {
+                "feature_extractors": [
+                    feature_extractor.to_dict()
+                    for feature_extractor in self.feature_extractors
+                ]
+            }
+        }
+
+    def feature_names(self) -> List[str]:
+        """
+        List of the feature names for the model.
+        """
+        if not hasattr(self, "_feature_names"):
```

Comment:
we seem to reuse this patern in multiple methods:

```
if not hasattr(self, attr_name)
  # assign attribute
return attribute
```
this seems like a good candidate for using the `cached_property` decorator: https://docs.python.org/3/library/functools.html#functools.cached_property

I see from the README that Eland supports python 3.8, 3.9, 3.10. So `cached_property` should be available.

https://github.com/elastic/eland/blob/926f0b9b5c5980e42db9e6e119a290a2eb4c7a60/README.md?plain=1#L57

Reply:
I did change the implementation to use cached_property. Thank you for the tip.